File: phase_velocity_interacting.jl
using LinearAlgebra
using Base
using JLD2
using MPSKitModels, TensorKit, MPSKit
using Statistics
using Plots
# using LsqFit
using Polynomials

# Define the function to compute residuals
function compute_residuals(fitted_curve, x_values, y_values)
    fitted_values = [fitted_curve(x) for x in x_values]
    residuals = y_values - fitted_values
    return residuals
end


am_tilde_0 = 0.3
Delta_g = -0.15
v = 0.0
bounds_k = pi/2
trunc = 4.0


name = "SanderDM_Thesis_2324/Dispersion_Delta_m_$(am_tilde_0)_delta_g_$(Delta_g)_v_$(v)_trunc_$(trunc)_all_sectors_newv"

@load name gs_energy energies Bs anti_energies anti_Bs

energies .+= Delta_g
anti_energies .-= Delta_g

k_values = LinRange(-bounds_k, bounds_k,length(energies))

# energies = [sqrt(am_tilde_0^2 + sin(k/2)^2) for k in 2*k_values]

order = 20

fitted_curve = fit(2*k_values, real.(energies)[:], order)

# Compute the derivative of the fitted curve
fitted_derivative = derivative(fitted_curve, 1)

# Define the range of k-values for which you want to compute the derivative
k_values_for_derivative = LinRange(-bounds_k, bounds_k, 1000)  # Adjust the number of points as needed

# Compute the values of the derivative for the specified range
derivative_values = [fitted_derivative(k) for k in k_values_for_derivative]

# Calculate the standard deviation of the derivative values
std_dev_derivative = std(derivative_values)

println("Standard deviation of derivative: $std_dev_derivative")

deriv = derivative(fitted, 1)

println("Velocity at k = $(k) is $(deriv(k))")


plt = scatter(2*k_values, real.(energies)[:], markerstrokewidth=0, label="Data")
plot!(fitted, extrema(2*k_values)..., label="Interpolation")
display(plt)

k_values_finer = LinRange(-bounds_k, bounds_k, 1000)
max_value, index = findmax([deriv(k) for k in k_values_finer])

using Optim


# Initial guess for the parameter
x0 = [0.0]  # Initial guess as an array

# Define the function to optimize
function objective_function(params)
    return deriv(params[1])
end

# Optimize the parameter to minimize f(x)
result = optimize(objective_function, x0, LBFGS())

# Extract the optimized parameter
optimized_parameter = result.minimizer

# Minimal value of f(x)
minimal_value = result.minimum

println("Optimized parameter: ", optimized_parameter)
println("Minimal value of f(x): ", 2*minimal_value)


using CSV
CSV.write("SanderDM_Thesis_2324/Derivative_info_Dispersion_Delta_m_$(am_tilde_0)_delta_g_$(Delta_g)_v_$(v)_trunc_$(trunc)_all_sectors_newv.csv", [(real.(energies)[:])], header=false)

using DelimitedFiles
open("SanderDM_Thesis_2324/Derivative_info_Dispersion_Delta_m_$(am_tilde_0)_delta_g_$(Delta_g)_v_$(v)_trunc_$(trunc)_all_sectors_newv.csv", "w") do io
    writedlm(io, (real.(energies)[:]), ',')
end

File: phase_velocity_interacting.py
import csv
import numpy as np
from scipy.optimize import curve_fit
import matplotlib.pyplot as plt

am_tilde_0 = 0.3
Delta_g = -0.15
v = 0.0
trunc = 4.0
bounds_k = np.pi/2

file = f"SanderDM_Thesis_2324/data/Derivative_info_Dispersion_Delta_m_{am_tilde_0}_delta_g_{Delta_g}_v_{v}_trunc_{trunc}_all_sectors_newv_more.csv"

energies = []

with open(file, newline='') as csvfile:
    # Create a CSV reader object
    csvreader = csv.reader(csvfile)
    
    # Iterate over each row in the CSV file
    for row in csvreader:
        # Each row is a list of strings representing the values in that row
        # Print the row
        energies.append(float(row[0]))

k_values = np.linspace(-bounds_k, bounds_k,len(energies))

def polynomial(x, *params):
    summation = 0
    for (i,param) in enumerate(params):
        summation += param * x**i
    return summation

def polynomial_derivative(x, *params):
    summation = 0
    for (i,param) in enumerate(params):
        summation += param * i * x**(i-1)
    return summation

order = 16
p0 = [1] * order
popt, pcov = curve_fit(polynomial, k_values, energies, p0 = p0)#, method = 'trf')

k_values_finer = np.linspace(-bounds_k,bounds_k, 1000)

split = True
if split:
    plt.plot(2*k_values, [e - Delta_g for e in energies], label = 'U(1) sector +1')
    plt.plot(2*k_values, [e + Delta_g for e in energies], label = 'U(1) sector -1')
else:
    plt.plot(2*k_values_finer, [polynomial(x, *popt) for x in k_values_finer], label = 'polynomial fit', c = 'red')
    plt.scatter(2*k_values, energies, label = 'data', s = 3)

plt.xlabel(r"momentum $k$", fontsize = 15)
plt.ylabel("energy", fontsize = 15)
plt.legend(fontsize=12)
plt.show()

k = 0.5

def get_data(k, popt, pcov, order):
    vel = polynomial_derivative(k, *popt)

    param_plus = [popt[i]+pcov[i,i] for i in range(order)]
    param_min = [popt[i]-pcov[i,i] for i in range(order)]

    vel_plus = polynomial_derivative(k, *param_plus) 
    vel_min = polynomial_derivative(k, *param_min) 

    print(f"For k = {k}: derivative is {vel/2}, max is {vel_plus/2}, min is {vel_min/2}")

    print(f"For k = {k}: std is {(vel_plus-vel)/2} or {(vel-vel_min)/2}")

get_data(0.5, popt, pcov, order)
get_data(-0.75, popt, pcov, order)
